### PR TITLE
Conditonal volumes, service and sidecar.

### PR DESCRIPTION
**Pull request by chunter0**

### CODE REVIEW

File: codegen/model/chart.go [hunk 1]
```diff
@@ -88,6 +88,11 @@ type Operator struct {
 	// the name of the other operator can be included in this list. This operator
 	// will not be provisioned unless both are enabled (by having values.enabled = true)
 	EnabledDependsOn []string
+
+	// (Optional) Use this field to specify if this operator should be deployed as
+	// a sidecar of an existing operator. When this field is set, the RBAC rules
+	// will be merged as well as Deployment fields such as volumes and volume mounts.
+	Operator string
```
Comment (chunter0) [inline codegen/model/chart.go:95]: ParentOperator

File: codegen/render/funcs.go [hunk 2]
```diff
@@ -131,6 +131,22 @@ func makeTemplateFuncs(customFuncs template.FuncMap) template.FuncMap {
 		"containerConfigs": containerConfigs,
 
 		"opVar": opVar,
+
+		"operatorSidecars": func(operators []model.Operator) (out map[string][]model.Operator) {
+			for _, operator := range operators {
+				// Skip operators that specify a sidecar
```
Comment (chunter0) [inline codegen/render/funcs.go:137]: // Skip operators that dont specify a sidecar

File: codegen/cmd_test.go [hunk 3]
```diff
@@ -40,6 +41,89 @@ var _ = Describe("Cmd", func() {
...
+						},
+						// Indicate this operator should be deployed as a sidecar
+						// of the glooMgmtServer operator
+						ParentOperator: "glooMgmtServer",
+					},
+				},
+			},
+		}
+
+		if err := cmd.Execute(); err != nil {
```
Comment (chunter0) [inline codegen/cmd_test.go:121]: I tested these changes via `go get` the commit directly into gloo-mesh-enterprise, I figure we can also have a unit test here but not sure where it should be placed

@josh-pritchard

File: codegen/templates/chart/operator-deployment.yamltmpl [hunk 4]
```diff
@@ -81,9 +78,23 @@ spec:
 [[- if $volumes ]]
       volumes:
 [[ toYaml $volumes | indent 6 ]]
+[[- end ]]
+[[- range $container := $containers -]]
+[[- range $volume := $container.Volumes ]]
+[[- if (gt (len $container.EnableStatement) 0) ]]
```
Comment (ilackarms) [inline codegen/templates/chart/operator-deployment.yamltmpl:84]: nit: you could swap this line with th one above it, reads a bit cleaner imo

File: codegen/templates/chart/operator-deployment.yamltmpl [hunk 5]
```diff
@@ -237,6 +251,40 @@ spec:
...
 
----
-
-[[ end ]]
\ No newline at end of file
+[[- range $container := $containers -]]
+[[- if and (gt (len $container.EnableStatement) 0) (gt (len $operator.Service.Ports) 0) ]]
+[[ printf "\n---" ]]
+[[ $containerVar := printf "$.Values.%s" (lower_camel $container.Name) ]]
+[[- if (gt (len $container.EnableStatement) 0) ]]
```
Comment (ilackarms) [inline codegen/templates/chart/operator-deployment.yamltmpl:258]: nit: we've already evaluated this to be true for this block

**Review by ilackarms — APPROVED** [review]

File: codegen/cmd_test.go [hunk 6]
```diff
@@ -40,7 +41,103 @@ var _ = Describe("Cmd", func() {
...
+						CustomEnableCondition: `and ($.Values.glooAgent.enabled) (not $.Values.glooAgent.runAsSidecar)`,
+					},
+				},
+			},
+			ManifestRoot: "codegen/test/chart",
+		}
+
+		Expect(cmd.Execute()).NotTo(HaveOccurred(), "failed to execute command")
+
+		absPath, err := filepath.Abs("./test/chart/templates/deployment.yaml")
```
Comment (josh-pritchard) [inline codegen/cmd_test.go:131]: This path is used in another test, We are not really seeing these templates for review here as they are getting overwritten by a later test

File: codegen/templates/chart/operator-deployment.yamltmpl [hunk 7]
```diff
@@ -237,6 +249,36 @@ spec:
...
+  type: {{ [[ $containerVar ]].serviceType }}
+  ports:
+  [[- range $port := $container.Service.Ports ]]
+  - name: [[ lower $port.Name ]]
+    port: {{ [[ $containerVar ]].ports.[[ $port.Name ]] }}
+  [[- end ]]
+{{- end }}
+[[- end ]]
+[[- end ]]
+[[ end ]]
```
Comment (josh-pritchard) [inline codegen/templates/chart/operator-deployment.yamltmpl:284]: This seems like a redundant Service, your test doesn't seem to reference the container.Service anyways so I am not sure what this is for, was this intended to be the sidecar?

**Review by josh-pritchard — COMMENTED** [review]
This implementation seems to have changed a bit from when we originally chatted with @Sodman. Instead of using the values from a completely different operator now we appear to use the container sidecar config. I am okay with this but it does lead to some redundant configuration. I also think the values are now not being completely handled correctly with this approach as they will get generated for the glooAgent block and the sidecar but only the glooAgent block is referenced. Can I see an example of this run against GME in a draft PR? It's easy for me to confirm it’s all correct that way.

**Review by josh-pritchard — APPROVED** [review]
LGTM

**Review by ilackarms — APPROVED** [review]

**Pull request merged**